File: RecordWorkouts/views/DatePickerController.m
#import "DatePickerController.h"
#import "../WRFormat.h"

@implementation DatePickerController {
    void (^_callbackHandler)(NSDate *date);
    UIDatePicker *picker;
    UITextField *_field;
    DatePickerController *_instance;
}

- (id) init: (UITextField *)field
toolbarBuilder: (ToolbarBuilder *)toolbarBuilder
   callback: (void (^)(NSDate *))callback; {
    self = [super init];
    _instance = self;
    _field = field;
    _callbackHandler = callback;
    
    picker = [[UIDatePicker alloc] init];
    [picker addTarget:_instance action:@selector(updateNewDate:) forControlEvents:UIControlEventValueChanged];
    picker.maximumDate = [NSDate date];
    
    _field.tintColor = [UIColor clearColor];
    [_field setInputAccessoryView:[self createToolbar:toolbarBuilder]];
    [_field setInputView:picker];
    return self;
}

- (UIToolbar*) createToolbar: (ToolbarBuilder *)toolbarBuilder {
    [toolbarBuilder addActionButton:@"Now" target:self action:@selector(setDateNow)];
    [toolbarBuilder addButton:toolbarBuilder.spacer];
    [toolbarBuilder addButton:toolbarBuilder.doneButton];
    return [toolbarBuilder create];
}

 - (void) setDateNow {
     [self setNewDate:[NSDate date]];
 }

- (void) updateNewDate: (UIDatePicker *) sender {
    [self setNewDate:sender.date];
}

- (void) setNewDate: (NSDate *) date {
    _field.text = [WRFormat formatDate:date];
    [picker setDate:date];
    _callbackHandler(date);
}

@end

File: workout-records/views/DatePickerController.swift
import UIKit

class DatePickerController {
    private var callbackHandler: ((_ date: Date) -> Void)
    private var picker: UIDatePicker
    private var field: UITextField
//    private var instance: DatePickerController
    
    init(_ field: UITextField, _ toolbarBuilder: ToolbarBuilder, callback: @escaping (Date) -> Void) {
//        instance = self
        self.field = field
        callbackHandler = callback
        
        picker = UIDatePicker()
//        picker.addTarget(instance, action: #selector(DatePickerController.updateNewDate(_:)), for: .valueChanged)
        picker.addTarget(self, action: #selector(updateNewDate(_:)), for: .valueChanged)
        
        self.field.tintColor = UIColor.clear
        self.field.inputAccessoryView = createToolbar(toolbarBuilder)
        self.field.inputView = picker
        setDateNow()
    }
    
    private func createToolbar(_ toolbarBuilder: ToolbarBuilder) -> UIToolbar {
        toolbarBuilder.addActionButton("Now", target: self, action: #selector(setDateNow))
        toolbarBuilder.addButton(toolbarBuilder.spacer)
        toolbarBuilder.addButton(toolbarBuilder.doneButton)
        return toolbarBuilder.create()
    }
    
    @objc func setDateNow() {
        let now = Date()
        picker.maximumDate = now
        setNewDate(now)
    }
    
    @objc private func updateNewDate(_ sender: UIDatePicker) {
        setNewDate(sender.date)
    }
    
    private func setNewDate(_ date: Date) {
        field.text = WRFormat.formatDate(date)
        picker.date = date
        callbackHandler(date)
    }
}
